contracts/staking/StakingMain.sol

```diff
@@ -0,0 +1,179 @@
+import "./interfaces/IStakingEvents.sol";
+import "./common/StakingConstants.sol";
+import "./common/StakingErrors.sol";
+import "./common/StakingStorage.sol";
+import "../common/ownership/Ownable.sol";
+import "../common/interfaces/IERC20.sol";
+import "../voucher/interfaces/IVoucher.sol";
+
+contract StakingMain is IStakingMain, IStakingEvents, Ownable, StakingStorage {
+    function stakeVotes(uint256 votesNumber) external override {
+        uint256 remainder = votesNumber - (votesNumber / StakingConstants.STAKE_STEP) * StakingConstants.STAKE_STEP;
+        if (votesNumber < StakingConstants.STAKE_STEP || remainder != 0) revert StakingErrors.WrongMultiplicity();
+
+        // the first stake must be more then the threshold
+        Stake[] storage stakes = _stakes[msg.sender];
+        if (stakes.length == 0 && votesNumber < _votesThreshold) revert StakingErrors.NotEnoughVotesForStake();
+
+        // check delay between stakes
+        if (stakes.length != 0) {
+            Stake memory lastStake = stakes[stakes.length - 1];
+            if (lastStake.startDate > block.timestamp) revert StakingErrors.StakeDateBeforeNow();
+
+            uint256 timeDelta = block.timestamp - lastStake.startDate;
+            if (timeDelta < StakingConstants.STAKE_COOLDOWN) revert StakingErrors.TooOftenStaking();
+        }
+
+        // check staker balance
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(msg.sender);
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        // append a new stake
+        _pushStake(msg.sender, block.timestamp, votesNumber);
+
+        // TODO: change IERC20 to IVote
+        IERC20(_config.voteAddress).transferFrom(msg.sender, address(this), votesNumber);
+        emit VotesStaked(block.timestamp, votesNumber);
+    }
+
+    function unstakeVotes() external override {
+        Stake[] storage stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        // remove stakes
+        uint256 vouchersNumber;
+        uint256 votesNumber;
+        while (stakes.length > 0) {
+            Stake memory stake = stakes[stakes.length - 1];
+            vouchersNumber += _calculateVouchers(stake);
+            votesNumber += stake.amount;
+
+            _popStake(msg.sender);
+        }
+
+        // ???
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(address(this));
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        IERC20(_config.voteAddress).transfer(msg.sender, votesNumber);
+        emit VotesUnstaked(msg.sender, votesNumber);
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(msg.sender, vouchersNumber);
+        emit VouchersMinted(msg.sender, vouchersNumber);
+    }
+
+    function withdrawVouchersForAll() external override isOwner {
+        for (uint256 i = 0; i < _stakers.length; i++) {
```

Comment:
Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption